app/services/visited_but_no_bp_service.rb

```diff
@@ -0,0 +1,54 @@
+class VisitedButNoBPService
+  def initialize(region, periods:)
+    @region = region
+    @periods = periods
+    @facilities = region.facilities
+  end
+
+  attr_reader :facilities
+  attr_reader :periods
+  attr_reader :region
+
+  def call
+    periods.each_with_object({}) do |period, result|
+      result[period] = patients_visited_with_no_bp_taken(period).count
+    end
+  end
+
+  def patients_visited_with_no_bp_taken(period)
+    Patient
+      .distinct
+      .with_hypertension
+      .where(registration_facility: facilities)
+      .joins(left_outer_join("appointments", period))
+      .joins(left_outer_join("prescription_drugs", period))
+      .joins(left_outer_join("blood_sugars", period, field: "recorded_at"))
+      .where("appointments.id IS NOT NULL OR prescription_drugs.id IS NOT NULL OR blood_sugars.id IS NOT NULL")
+      .where(no_blood_pressures_recorded(period))
+  end
+
+  def between_clause(table, field, period)
+    begin_date = period.advance(months: -3).start_date
+    end_date = period.end_date
+    sql = ["#{table}.#{field} >= ? AND #{table}.#{field} <= ?", begin_date, end_date]
+    ActiveRecord::Base.sanitize_sql_for_conditions(sql)
+  end
+
+  def no_blood_pressures_recorded(period)
+    <<-SQL
+      NOT EXISTS (
+        SELECT 1
+        FROM blood_pressures bps
+        WHERE patients.id = bps.patient_id
+          AND #{between_clause("bps", "recorded_at", period)}
+      )
+    SQL
+  end
```

Comment:
☝️

Reply:
This looks correct, but I don't understand where `patients` is appearing from? Raw SQL is very hard to understand when it's partially applied. Is this basically,

```
BloodPressure.where.not(patient: patients, recorded_at: begin_date...end_date).exists?
```

If so, 

Could we re-write `patients_visited_with_no_bp_taken` with a normal `left_outer_joins` instead?

```
def between(period)
  begin_date = period.advance(months: -3).start_date
  end_date = period.end_date
  Range.new(begin_date, end_date)
end

def patients_visited_with_no_bp_taken(period)
  Patient
    .distinct
    .with_hypertension
    .where(registration_facility: facilities)
    .left_outer_joins(:appointments).where(appointments: {recorded_at: between(period)})
    .left_outer_joins(:prescription_drugs).where(prescription_drugs: {recorded_at: between(period)})
    .left_outer_joins(:blood_sugars).where(blood_sugars: {recorded_at: between(period)})
    .left_outer_joins(:blood_pressures).where(blood_pressures: {id: nil})
    .where("appointments.id IS NOT NULL OR prescription_drugs.id IS NOT NULL OR blood_sugars.id IS NOT NULL")
end
```

In the future, this may compose a lot better.

Reply:
I dont' think any of these ActiveRecord based approaches are really clear, to be honest 😄 . I think these kind of queries are really straining the abilities of our ORM, and would rather we just drop into SQL....especially because this should really be **one** query, instead of 24 queries repeated for each period.

I'm going to look into getting some of these queries rolled up into one via the Postgres [generate_series](https://www.postgresql.org/docs/10/functions-srf.html) functions, which will make these queries longer but should be able to get all our data in one trip.  When I do that I'll probably try it out with [GitHub::SQL](https://github.com/github/github-ds) to see how that works out.

Reply:
I don't mind dropping into SQL at all. I have used query builders (like the one you're suggesting) in the past and even [written](https://github.com/nilenso/honeysql-postgres) one and my general feeling is that they lie under the pretense of an abstraction but aren't really actually one (you can't compose relational algebra using strings, you lose the mapping to `ActiveRecord`). However, they are far better than raw SQL which generally ends up leaking subtle bugs. I also agree if it's performant and not a terribly complicated query to write raw, it's better to write a single one. If you're looking to migrate to that, I'm 👍.

However, our queries are operating on well-defined relations (and not flat / arbitrary table structures) and we can probably write ~95% of them easily with AR. Writing with AR has many advantages (and some disadvantages as well):

* They compose better
* They are easier to drive through the REPL
* Atomic pieces can be tested in isolation
* Useful default scopes are pre-applied
* The SQL building noise is hidden away from application code
* Lesser need to munge / interpolate strings awkwardly

Disadvantages:

* Can't express everything (postgres is vast and not everything is supported)
* The queries may not always be performant
* Some scopes in AR can bite you, so you have to be a diligent 

That said, writing SQL can often "feel" a lot better because, you can have one giant SQL to look at that does all the things that you think it should do and you operate within that space. For complex reporting queries that need to be performant this may very well be a good choice (I just haven't encountered a scenario to justify it yet other than the materialized views).

Specifically about the refactor,

I'm sure you can tweak the name of `scope_visits_by` to suggest it's `left_outer_join` intention clearly, my basic point was to avoid writing things like this,

```
   <<-SQL
      NOT EXISTS (
        SELECT 1
        FROM blood_pressures bps
        WHERE patients.id = bps.patient_id
          AND #{between_clause("bps", "recorded_at", period)}
      )
    SQL
```

Which is an invalid query by itself and isn't very clear that it only works when applied to specific parts of another larger SQL, which in comparison this,

```
  Patient.left_outer_joins(model).where(model => {time_field => period})
```

...is not. It's both testable and runnable in isolation.

And secondly, to not lose the original point 😛 , this was my main suggestion to arrive at the `no_bp` query,

```
  Patient.left_outer_joins(:blood_pressures).where(blood_pressures: {recorded_at: period, id: nil})
```

Just my 2c.